src/main/java/org/opentripplanner/updater/package.md

```diff
@@ -0,0 +1,104 @@
+
+## Realtime Concurrency
+
+The following approach to realtime concurrency was devised around 2013 when OTP first started consuming realtime data that affected routing results rather than just displaying messages. At first, the whole realtime system followed this approach. Some aspects of this system were maintained in subsequent work over the years, but because the details and rationale were not fully documented, misinterpretations and subtle inconsistencies were introduced.
+
+On 11 January 2024 a team of OTP developers reviewed this realtime concurrency approach together. The conclusion was that this approach remains sound, and that any consistency problems were not due to the approach itself, but rather due to its partial or erroneous implementation in realtime updater classes. Therefore, we decided to continue applying this approach in any new work on the realtime subsystem, at least until we encounter some situation that does not fit within this model. All existing realtime code that is not consistent with this approach should progressively be brought in line with it. 
+
+The following is a sequence diagram showing how threads are intended to communicate. Unlike some common forms of sequence diagrams, time is on the horizontal axis here. Each horizontal line represents either a thread of execution (handling incoming realtime messages or routing requests) or a queue or buffer data structure. Dotted lines represent object references being handed off, and solid lines represent data being copied.
+
+![Realtime sequence diagram](images/updater-threads-queues.svg)
+
+At the top of the diagram are the GraphUpdater implementations. These fall broadly into two categories: polling updaters and streaming updaters. Polling updaters periodically send a request to server (often just a simple HTTP server) which returns a file containing the latest version of the updates. Streaming updaters are generally built around libraries implementing message-oriented protocols such as AMQP or WebSockets, which fire a callback each time a new message is received. Polling updaters tend to return a full dataset describing the entire system state on each polling operation, while streaming updaters tend to receive incremental messages targeting individual transit trips. As such, polling updaters execute relatively infrequently (perhaps every minute or two) and process large responses, while streaming updaters execute very frequently (often many times per second) and operate on small messages in short bursts. Polling updaters are simpler in many ways and make use of common HTTP server components, but they introduce significant latency and redundant communication. Streaming updaters require more purpose-built or custom-configured components including message brokers, but bandwidth consumption and latency are lower, allowing routing results to reflect vehicle delays and positions immediately after they're reported.
```

Comment:
While AMQP and Websockets _are_ message oriented protocols, I would probably mention MQTT, for which we have an implementation, as opposed to the other two.

Reply:
It looks like the AbstractAzureSiriUpdater (and its ET and SX subclasses) are AMQP-based. Although these are sandbox extensions, the purpose of this package documentation is to describe the general framework (message-oriented updates where the protocol is up to the implementer or hosting environment) so it seems reasonable to mention a couple of example protocols including AMQP. That said, point taken that the only non-sandbox message oriented updater we have is MQTT based, so MQTT should be mentioned most prominently, and we no longer have any Websocket-based updater.